agents/aggregator/aggregator_agent.py

```diff
@@ -18,463 +13,511 @@ class Provider:
-        self.log.info("Ending file")
+        self.log.info("Aggregator running...")
+        self.hksess.start_time = time.time()
+        self.hksess.session_id = self.generate_id()
 
-        for pid, prov in self.providers.items():
-            if prov.frame_start_time is None:
-                continue
+        self.rotator.Process(self.hksess.session_frame())
 
-            with prov.lock:
-                frame = prov.to_frame(self.hksess)
-                prov.clear()
+        self.aggregate = True
+        while self.aggregate:
+            time.sleep(self._loop_time)
 
-            self.writer(frame)
+            self.remove_stale_providers()
+            self.write_providers_to_queue()
+            self.write_queue_to_disk()
 
-        self.writer(core.G3Frame(core.G3FrameType.EndProcessing))
-        self.writer = None
+            self.rotator.flush()
 
-        return
+        self.write_providers_to_queue(write_all=True)
+        self.write_queue_to_disk()
 
-    def start_aggregate(self, session, params={}):
-        """
-        PROCESS: Starts the aggregation process.
+        self.rotator.close_file()
 
-        Args:
-            time_per_file (int, optional):
-                Specifies how much time should elapse before starting a new
-                file (in seconds). Defaults to 1 hr.
+        return True
 
-            time_per_frame (int, optional):
-                Specifies how much time should elapse before starting a new
-                frame (in seconds). Defaults to 10 minutes.
+    def stop(self):
+        """Signals for aggregation to stop."""
+        self.aggregate = False
 
-            data_dir (string, optional):
-                Path of directory to store data. Defaults to 'data/'.
-        """
-        if params is None:
-            params = {}
 
-        time_per_file = params.get("time_per_file", self.time_per_file) # [s]
-        data_dir = params.get("data_dir", self.data_dir)
+class AggregatorAgent:
+    def __init__(self, agent, args):
+        self.agent: ocs_agent.OCSAgent = agent
+        self.log = agent.log
 
-        self.log.info("Starting data aggregation in directory {}".format(data_dir))
-        session.set_status('running')
+        self.aggregator = Aggregator(args, self.log)
 
-        self.hksess.start_time = time.time()
-        # Encode a suitable agg_session_id.  if 1:
-        elements = [(int(self.hksess.start_time), 32),
-                    (os.getpid(), 14),
-                    (binascii.crc32(bytes(self.hksess.description, 'utf8')), 14)]
-        agg_session_id = 0
-        for i, b in elements:
-            agg_session_id = (agg_session_id << b) | (i % (1<<b))
-        self.log.info("New aggregator session_id=%i" % agg_session_id)
-        # The cast here to G3Int can be dropped in the future -- spt3g
-        # was updated on May 1 2019 to handle direct puts of int64_t.
-        self.hksess.session_id = core.G3Int(agg_session_id)
+        # SUBSCRIBES TO ALL FEEDS!!!!
```

Comment:
Indeed, this is a significant change.  Have you taken a census of what Agents currently use non-recorded feeds?  I'm not against having separate domains for "recordable" and "ephemeral" data streams.  (Certainly for now it should be the ephemeral ones that move elsewhere.)

Reply:
I think most other "ephemeral" feeds are heartbeats (its arguable whether they should fall under 'feeds' in the first place). I think currently, status messages don't actually match this subscription, because they're published to `<agent_address>.feed`. The registry has the `agent_activity` feed that's not recorded, but that doesn't really need to exist after this PR... Just did a quick search of ocs and socs, and it actually doesn't look like there are any other non-aggregated feeds.